File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxDataBinding.java
package xyz.truenight.databinding.rxjava;

import android.databinding.BaseObservable;
import android.databinding.ObservableBoolean;
import android.databinding.ObservableByte;
import android.databinding.ObservableChar;
import android.databinding.ObservableDouble;
import android.databinding.ObservableField;
import android.databinding.ObservableFloat;
import android.databinding.ObservableInt;
import android.databinding.ObservableList;
import android.databinding.ObservableLong;
import android.databinding.ObservableMap;
import android.databinding.ObservableShort;
import android.support.annotation.IntDef;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import io.reactivex.Observable;
import io.reactivex.ObservableTransformer;
import xyz.truenight.utils.Optional;

/**
 * Copyright (C) 2017 Mikhail Frolov
 */

public class RxDataBinding {
    private RxDataBinding() {
    }

    public static Observable<Boolean> asObservable(ObservableBoolean field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field.get());
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field.get());
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static Observable<Integer> asObservable(ObservableInt field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field.get());
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field.get());
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static Observable<Long> asObservable(ObservableLong field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field.get());
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field.get());
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static Observable<Short> asObservable(ObservableShort field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field.get());
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field.get());
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static Observable<Float> asObservable(ObservableFloat field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field.get());
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field.get());
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static Observable<Double> asObservable(ObservableDouble field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field.get());
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field.get());
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static Observable<Character> asObservable(ObservableChar field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field.get());
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field.get());
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static Observable<Byte> asObservable(ObservableByte field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field.get());
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field.get());
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static <T> Observable<Optional<T>> asObservable(ObservableField<T> field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(Optional.ofNullable(field.get()));
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(Optional.ofNullable(field.get()));
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static <T extends BaseObservable> Observable<T> asObservable(T field) {
        return Observable.create(emitter -> {
            android.databinding.Observable.OnPropertyChangedCallback callback = new android.databinding.Observable.OnPropertyChangedCallback() {
                @Override
                public void onPropertyChanged(android.databinding.Observable sender, int propertyId) {
                    emitter.onNext(field);
                }
            };
            emitter.setCancellable(() -> field.removeOnPropertyChangedCallback(callback));
            emitter.onNext(field);
            field.addOnPropertyChangedCallback(callback);
        });
    }

    public static class ListChange<T> {
        public static final int ALL = 0;
        public static final int CHANGED = 1;
        public static final int INSERTED = 2;
        public static final int MOVED = 3;
        public static final int REMOVED = 4;

        private ObservableList<T> list;
        private int action;
        public int start;
        public int count;
        public int to;

        public ObservableList<T> getList() {
            return list;
        }

        @Action
        public int getAction() {
            return action;
        }

        public boolean isAction(@Action int action) {
            return this.action == action;
        }

        public int getStart() {
            return start;
        }

        public int getCount() {
            return count;
        }

        public int getTo() {
            return to;
        }

        public ListChange(ObservableList<T> ts) {
            list = ts;
        }

        public static <T> ListChange<T> changed(ObservableList<T> ts, int start, int count) {
            ListChange<T> change = new ListChange<>(ts);
            change.action = CHANGED;
            change.start = start;
            change.count = count;
            return change;
        }

        public static <T> ListChange<T> moved(ObservableList<T> ts, int start, int to, int count) {
            ListChange<T> change = new ListChange<>(ts);
            change.action = MOVED;
            change.start = start;
            change.to = to;
            change.count = count;
            return change;
        }

        public static <T> ListChange<T> inserted(ObservableList<T> ts, int start, int count) {
            ListChange<T> change = new ListChange<>(ts);
            change.action = INSERTED;
            change.start = start;
            change.count = count;
            return change;
        }

        public static <T> ListChange<T> removed(ObservableList<T> ts, int start, int count) {
            ListChange<T> change = new ListChange<>(ts);
            change.action = REMOVED;
            change.start = start;
            change.count = count;
            return change;
        }

        @IntDef({ALL, CHANGED, INSERTED, MOVED, REMOVED})
        @Retention(RetentionPolicy.SOURCE)
        public @interface Action {

        }
    }

    public static <T> Observable<ListChange<T>> asObservable(ObservableList<T> list) {
        return Observable.create(emitter -> {
            ObservableList.OnListChangedCallback<ObservableList<T>> callback = new ObservableList.OnListChangedCallback<ObservableList<T>>() {
                @Override
                public void onChanged(ObservableList<T> ts) {
                    next(new ListChange<>(ts));
                }

                @Override
                public void onItemRangeChanged(ObservableList<T> ts, int i, int i1) {
                    next(ListChange.changed(ts, i, i1));
                }

                @Override
                public void onItemRangeInserted(ObservableList<T> ts, int i, int i1) {
                    next(ListChange.inserted(ts, i, i1));
                }

                @Override
                public void onItemRangeMoved(ObservableList<T> ts, int i, int i1, int i2) {
                    next(ListChange.moved(ts, i, i1, i2));
                }

                @Override
                public void onItemRangeRemoved(ObservableList<T> ts, int i, int i1) {
                    next(ListChange.removed(ts, i, i1));
                }

                private void next(ListChange<T> change) {
                    emitter.onNext(change);
                }
            };
            emitter.setCancellable(() -> list.removeOnListChangedCallback(callback));
            emitter.onNext(new ListChange<>(list));
            list.addOnListChangedCallback(callback);
        });
    }

    public static <K, V> Observable<ObservableMap<K, V>> asObservable(ObservableMap<K, V> field) {
        return Observable.create(emitter -> {
            ObservableMap.OnMapChangedCallback<ObservableMap<K, V>, K, V> callback = new ObservableMap.OnMapChangedCallback<ObservableMap<K, V>, K, V>() {
                @Override
                public void onMapChanged(ObservableMap<K, V> sender, K key) {
                    emitter.onNext(field);
                }
            };
            emitter.setCancellable(() -> {
                field.removeOnMapChangedCallback(callback);
            });
            emitter.onNext(field);
            field.addOnMapChangedCallback(callback);
        });
    }

    public static <T> ObservableTransformer<T, T> loadingNext(ObservableBoolean loading) {
        return observable -> observable
                .doOnSubscribe((d) -> loading.set(true))
                .doOnNext((t) -> loading.set(false));
    }

    public static <T> ObservableTransformer<T, T> error(ObservableBoolean error) {
        return observable -> observable
                .doOnNext((t) -> error.set(false))
                .doOnError((th) -> error.set(true));
    }

    public static <T> void setField(ObservableField<T> field, T data) {
        if (field.get() != data) {
            field.set(data);
        } else {
            field.notifyChange();
        }
    }

    public static void setField(ObservableBoolean field, boolean data) {
        if (field.get() != data) {
            field.set(data);
        } else {
            field.notifyChange();
        }
    }
}

File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxDataBinding.kt
@file:JvmName("RxDataBinding")

package xyz.truenight.databinding.rxjava


import androidx.databinding.*
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import xyz.truenight.utils.optional.Optional
import xyz.truenight.utils.optional.toOptional

/**
 * Copyright (C) 2017 Mikhail Frolov
 */

fun ObservableBoolean.asFlowable(): Flowable<Boolean> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun ObservableInt.asFlowable(): Flowable<Int> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun ObservableLong.asFlowable(): Flowable<Long> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun ObservableShort.asFlowable(): Flowable<Short> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun ObservableFloat.asFlowable(): Flowable<Float> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun ObservableDouble.asFlowable(): Flowable<Double> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun ObservableChar.asFlowable(): Flowable<Char> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun ObservableByte.asFlowable(): Flowable<Byte> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun <T : Any> ObservableField<T?>.asFlowable(): Flowable<Optional<T>> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(get().toOptional())
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(get().toOptional())
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

fun <T : BaseObservable> T.asFlowable(): Flowable<T> {
    return Flowable.create({ emitter ->
        val callback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {
                emitter.onNext(this@asFlowable)
            }
        }
        emitter.setCancellable { removeOnPropertyChangedCallback(callback) }
        emitter.onNext(this)
        addOnPropertyChangedCallback(callback)
    }, BackpressureStrategy.LATEST)
}

class ListChange<T> private constructor(val list: ObservableList<T>, val action: Action = Action.ALL, val start: Int = 0, val count: Int = 0, val to: Int = 0) {

    fun isAction(action: Action): Boolean {
        return this.action == action
    }

    companion object {

        internal fun <T> all(ts: ObservableList<T>) = ListChange(ts)

        internal fun <T> changed(ts: ObservableList<T>, start: Int, count: Int) =
                ListChange(
                        list = ts,
                        action = Action.CHANGED,
                        start = start,
                        count = count
                )

        internal fun <T> moved(ts: ObservableList<T>, start: Int, to: Int, count: Int) =
                ListChange(
                        list = ts,
                        action = Action.MOVED,
                        start = start,
                        to = to,
                        count = count
                )

        internal fun <T> inserted(ts: ObservableList<T>, start: Int, count: Int) =
                ListChange(
                        list = ts,
                        action = Action.INSERTED,
                        start = start,
                        count = count
                )

        internal fun <T> removed(ts: ObservableList<T>, start: Int, count: Int) =
                ListChange(
                        list = ts,
                        action = Action.REMOVED,
                        start = start,
                        count = count
                )
    }
}

enum class Action {
    ALL,
    CHANGED,
    INSERTED,
    MOVED,
    REMOVED
}

fun <T> ObservableList<T>.asFlowableChanges(): Flowable<ListChange<T>> = Flowable.create({ emitter ->
    val callback = object : ObservableList.OnListChangedCallback<ObservableList<T>>() {
        override fun onChanged(ts: ObservableList<T>) = next(ListChange.all(ts))

        override fun onItemRangeChanged(ts: ObservableList<T>, i: Int, i1: Int) =
                next(ListChange.changed(ts, i, i1))

        override fun onItemRangeInserted(ts: ObservableList<T>, i: Int, i1: Int) =
                next(ListChange.inserted(ts, i, i1))

        override fun onItemRangeMoved(ts: ObservableList<T>, i: Int, i1: Int, i2: Int) =
                next(ListChange.moved(ts, i, i1, i2))

        override fun onItemRangeRemoved(ts: ObservableList<T>, i: Int, i1: Int) =
                next(ListChange.removed(ts, i, i1))

        private fun next(change: ListChange<T>) = emitter.onNext(change)
    }
    emitter.setCancellable { removeOnListChangedCallback(callback) }
    emitter.onNext(ListChange.all(this))
    addOnListChangedCallback(callback)
}, BackpressureStrategy.LATEST)

fun <T> ObservableList<T>.asFlowable(): Flowable<ObservableList<T>> = Flowable.create({ emitter ->
    val callback = object : ObservableList.OnListChangedCallback<ObservableList<T>>() {
        override fun onChanged(ts: ObservableList<T>) = next(ts)

        override fun onItemRangeChanged(ts: ObservableList<T>, i: Int, i1: Int) = next(ts)

        override fun onItemRangeInserted(ts: ObservableList<T>, i: Int, i1: Int) = next(ts)

        override fun onItemRangeMoved(ts: ObservableList<T>, i: Int, i1: Int, i2: Int) = next(ts)

        override fun onItemRangeRemoved(ts: ObservableList<T>, i: Int, i1: Int) = next(ts)

        private fun next(change: ObservableList<T>) = emitter.onNext(change)
    }
    emitter.setCancellable { removeOnListChangedCallback(callback) }
    emitter.onNext(this)
    addOnListChangedCallback(callback)
}, BackpressureStrategy.LATEST)

fun <K, V> ObservableMap<K, V>.asFlowable(): Flowable<ObservableMap<K, V>> =
        Flowable.create({ emitter ->
            val callback = object : ObservableMap.OnMapChangedCallback<ObservableMap<K, V>, K, V>() {
                override fun onMapChanged(sender: ObservableMap<K, V>, key: K) {
                    emitter.onNext(this@asFlowable)
                }
            }
            emitter.setCancellable { removeOnMapChangedCallback(callback) }
            emitter.onNext(this)
            addOnMapChangedCallback(callback)
        }, BackpressureStrategy.LATEST)